bio/bio.go

```diff
@@ -0,0 +1,195 @@
+/*
+Package bio provides utilities for reading and writing sequence data.
+*/
+package bio
+
+import (
+	"bufio"
+	"errors"
+	"io"
+	"math"
+	"os"
+
+	"github.com/TimothyStiles/poly/bio/fasta"
+)
+
+type Format int
+
+const (
+	Fasta Format = iota
+	Fastq
+	Gff
+	Genbank
+	Slow5
+	Pileup
+	Uniprot
+	Rebase
+)
+
+// DefaultMaxLineLength variables are defined for performance reasons. While
+// parsing, reading byte-by-byte takes far, far longer than reading many bytes
+// into a buffer. In golang, this buffer in bufio is usually 64kb. However,
+// many files, especially slow5 files, can contain single lines that are much
+// longer. We use the default maxLineLength from bufio unless there is a
+// particular reason to use a different number.
+const defaultMaxLineLength int = bufio.MaxScanTokenSize // 64kB is a magic number often used by the Go stdlib for parsing.
+var DefaultMaxLengths = map[Format]int{
+	Fasta:   defaultMaxLineLength,
+	Fastq:   8 * 1024 * 1024, // The longest single nanopore sequencing read so far is 4Mb. A 8mb buffer should be large enough for any sequencing.
+	Gff:     defaultMaxLineLength,
+	Genbank: defaultMaxLineLength,
+	Slow5:   128 * 1024 * 1024, // 128mb is used because slow5 lines can be massive, since a single read can be many millions of base pairs.
+	Pileup:  defaultMaxLineLength,
+	Uniprot: defaultMaxLineLength,
+	Rebase:  defaultMaxLineLength,
+}
+
+/******************************************************************************
+Aug 30, 2023
+
+Lower level interfaces
+
+******************************************************************************/
+
+type LowLevelParser[DataType fasta.Record, DataTypeHeader fasta.Header] interface {
+	Header() (DataTypeHeader, int64, error)
+	Next() (DataType, int64, error)
+}
+
+/******************************************************************************
+
+Higher level parse
+
+******************************************************************************/
+
+type Parser[DataType fasta.Record, DataTypeHeader fasta.Header] struct {
+	LowLevelParser LowLevelParser[DataType, DataTypeHeader]
+}
+
+var emptyParser Parser[fasta.Record, fasta.Header] = Parser[fasta.Record, fasta.Header]{}
+
+func NewParser(format Format, r io.Reader) (*Parser[fasta.Record, fasta.Header], error) {
+	var maxLineLength int
+	switch format {
```

Comment:
Since we have a map of max lengths now, this switch isn't necessary.

Reply:
True, fixing!